packages/@lwc/engine-core/src/framework/components/dynamic.ts

```diff
@@ -0,0 +1,103 @@
+import { LightningElement, LightningElementConstructor } from '../base-lightning-element';
+import { ComponentDef, getComponentInternalDef } from '../def';
+import {
+    appendVM,
+    computeShadowMode,
+    createVMContext,
+    getAssociatedVMIfPresent,
+    hasWireAdapters,
+    resetComponentRoot,
+    runConnectedCallback,
+    runDisconnectedCallback,
+    VM,
+    VMState,
+} from '../vm';
+import { invokeComponentConstructor } from '../invoker';
+import { installWireAdapters } from '../wiring';
+
+function replaceDefinitionForVM(vm: VM, newDef: ComponentDef, replacedCallback: () => void) {
+    if (vm.state === VMState.connected) {
+        runDisconnectedCallback(vm);
+    }
+    if (vm.state === VMState.disconnected) {
+        // there is a possibility that the elm is disconnected, and the content
+        // inside its shadow is still visible to the new component, since the
+        // element is reused and runDisconnectedCallback doesn't remove its
+        // content, we need to do it manually.
+        resetComponentRoot(vm);
+    }
+    // remove content of the shadow if any and any internal cache
+    vm.def = newDef;
+    vm.context = createVMContext();
+    vm.cmpProps = create(null);
+    vm.cmpFields = create(null);
+    vm.refVNodes = null;
+
+    // validation of the mode... it must match the mode of the original implementation
+    // users can extend DynamicLightingElement to adjust the mode at will.
+    // do we need the same thing for cm.renderMode? I believe we do!
+    if (vm.shadowMode !== computeShadowMode(vm)) {
+        throw new TypeError(`Incompatible ShadowMode`);
+    }
+    // Create a new component instance associated to the vm and the element.
+    invokeComponentConstructor(vm, newDef.ctor);
+    // Initializing the wire decorator per instance only when really needed
+    if (hasWireAdapters(vm)) {
+        installWireAdapters(vm);
+    }
+
+    // what about existing props? should that be carry on by the diffing algo?
+    replacedCallback();
+
+    if (vm.elm.isConnected) {
+        // virtual re-insertion of the component into the DOM, by faking it.
+        runConnectedCallback(vm);
+        appendVM(vm);
+    }
+}
+
+function getVMFromComponent(obj: any): VM {
+    const vm = getAssociatedVMIfPresent(obj);
+    if (!vm) {
+        throw TypeError(`Invalid Invocation`);
+    }
+    return vm;
+}
+
+export class DynamicLightingElement extends LightningElement {
+    constructor() {
+        super();
+        let ctor: LightningElementConstructor | null = null;
+        const { replacedCallback } = this;
+        // installing the elm.ctor public property, it is not reflective to an attribute,
+        // and it is not configurable to avoid something messing with its public API. The
+        // fact that it is installed on instance guarantees that changing the proto-chain
+        // will not affect its behavior.
+        defineProperty(getVMFromComponent(this).elm, 'ctor', {
```

Comment:
Does this need to be a public property? Could we use a `WeakMap` instead?

Reply:
it can't be a weakmap if we want to allow `DynamicLightningElement.CustomElementConstructor` to be useful. Basically, if it is a weakmap, then a public API will have to be exposed from LWC so people can set the replace in motion by calling that API with an element and the new ctor, and that api will do what we do here... but then exporting a LWC as web components makes it a lot less useful. Another option is to make it a method rather than a setter, which I'm fine with as well.